### DISCUSSION TITLE
Ansible-like tool with Nu

### DISCUSSION THREAD

**Discussion by atoktoto** [nushell/nushell#11650] [Ideas]
Hey everyone, I've had an idea and I would like to show it to you to get your thoughts:

**An Ansible-like tool based on Nu**: you run a Nu script locally that SSHs into remote machines and runs some Nu scripts there to install and configure whatever is needed. To achieve idempotency things running on remote machines are checking first if the thing is already in place. 

For this to work, we would need:
- some local Nu code that collects inputs: files needed and remote machine addresses and such
- a `run-on` command that runs the provided block on each machine on the list using SSH connection
- a `resource` command that describes what needs to be done and how to check if it was done already
- parametric `resources` that can be reused to keep it DRY

I started drafting how the usage could look like. This example makes sure that mysql package is installed and some example file is present on all virtual machines tagged with `mysql` tag: 

```nushell
let vms = proxmox get inventory
let mysql_vms = $vms | where {|it| "mysql" in $it.tags }

let ctx = {
    config_text: (open "example.toml")
}

run-on $mysql_vms $ctx {

    # runs GET checks with TEST, if false runs SET and re-tests (GET, TEST). Returns both GETs
    # GETs can be cached
    # SETs do not have to be idempotent (assured by GET, TEST)
    resource "Upload config" {
        set: { cd "/example" ; $ctx.config_text | save "config.yaml" }
        get: { cd "/example" ; ls }
        test: { find "config.yaml" }
    }

    # reusable resources
    resource apt-install mysql-server
}
```

The second resource is an usage of reusable resource declaration:

```nushell
def "resource apt-install" [name: string] {
    resource "Something to install" {
        get: { dpkg -l | detect columns -s 3 | skip 1 | select Name Version }
        test: { |name| find $name }
        set: { |name| apt-get install $name }
    }
}
```

Similar to Ansible, this would only apply the changes if needed and on the subsequent runs would just verify that the state is as expected using `get` and `test` functions. If `get | test | is-empty` it means that we need to run `set`.

My thinking was that I might be able to implement `run-on` command as a Nu-plugin but I've since learned that blocks are only sent over as `BlockId` and without access to internals you cannot do anything with it. 

I also tried to stringify blocks using `view source`  but this creates confusing situations when using closures as I cannot know what was closed over. 

```nushell
def "resource" [name: string, gts] {
    let get = try { view source $gts.get }
    let test = try { view source $gts.test }
    let set = try { view source $gts.set }
    let serialized = { context: $gts.context, get: $get, test: $test, set: $set }
    echo $gts.context.config
}
```

Seems to me that Nu should have enough of expressive power to construct this and using a shell to do shell things would be a much better fit then using an combination of YAML and Python. Wdyt?

**Comment by hugosenari** [comment]
Nice idea, I would like that for my personal workflow.
My current thoughts are:
- How to make it run in every machine without a previous installation of an 'agent'
  - Currently I download the static binary to my local machine then I push it to target server
  - But would be good If we had a transpiller from nu to sh.
- How to transfer security sensitive data over ssh in the way it isn't in history, isn't in PS, etc.
  -  Currently I :see_no_evil: 
  - I have no idea how Ansible solves this.
  - But would be good if we all have TPM2 enabled and only destination server could decrypt the information or SPIFFE authentication.
- My ops team jumped the algorithm classes (like developers skip network classes), so a descriptive YAML should be easier for them.

You may need a lazy interpreter of nu script, or someway to get it AST than transfers it to the other machine.
Or embed nu scripts in nu scripts, but YAML has an advantage, since YAML|Nu has less conflict (needed to escape chars) than Nu|Nu.

Good luck!!

**Reply by ghost** [reply]
my thoughts on your thoughts:

> Or embed nu scripts in nu scripts

alternative: `view source`: it allows converting a nu closure back into its source-code.  
syntax highlight, etc works out of the box; no need for string-escaping, etc  
example: `let print_foo = {|| print "foo"}; view source $print_foo`

> But would be good If we had a transpiller from nu to sh.

i don't think that would be any easier. example code:
```nu
http get -r (
  http get foo.com/program/versions.json
  | get (uname -p)
  | get download_url
)
| save -r /usr/local/bin/program
```
this would require either a lot of awk, etc or jq at the very least. if we have to install something either way its probably not really worth it to transpile

> I have no idea how Ansible solves this.

<https://docs.ansible.com/automation-controller/latest/html/administration/secret_handling.html>

> My ops team jumped the algorithm classes (like developers skip network classes), so a descriptive YAML should be easier for them.

one good part about nu is: you can just use nuon or json and its valid nu code.
all the logic is optional.
```nu
[
  name: "install firefox"
  code: {|| apt install firefox -y}
  requirements: ["apt update"]
]
```
is about as readable / easy to learn as the following in my opinion and has the additional upside of working syntax highlighting, cross-line variable content, the option to add own logic, etc
```yaml
name: install firefox
code:
  - apt install firefox -y
requirements:
  - apt update
```
and it would also be possible to just transpile from yaml if it uses nu internally.

also i personally dislike yaml due to its annoying and weird extra features, which just cause bugs.
and im not sure if toml / ron / … would really be a better choice.
nu ansible would probably not be useful if its just ansible re implemented in nu.
it would be useful if we make full use of nu.

**Comment by theAkito** [comment]
If this idea still alive?
